auth0/auth0_client.go

```
@@ -4,18 +4,54 @@ import (
 	"bytes"
 	"encoding/json"
 	"fmt"
-
 	"github.com/parnurzeal/gorequest"
+	"net/http"
 )
 
 type AuthClient struct {
 	config *Config
 }
 
-func NewClient(config *Config) *AuthClient {
+func NewClient(clientId string, clientSecret string, config *Config) (*AuthClient, error) {
+
+	token, err := getToken(clientId, clientSecret, config.domain, config.apiUri)
+
+	if err != nil {
+		return nil, fmt.Errorf("auth0 provider init failed, error: %v", err)
+	}
+
+	config.accessToken = token
+
 	return &AuthClient{
 		config: config,
+	}, nil
+}
+
+func getToken(clientId string, clientSecret string, domain string, apiUri string) (string, error) {
+	auth0LoginRequest := &LoginRequest{
+		ClientId:     clientId,
+		ClientSecret: clientSecret,
+		Audience:     apiUri,
+		GrantType:    "client_credentials",
 	}
+
+	res, body, errs := gorequest.New().Post("https://" + domain + "/oauth/token").Send(auth0LoginRequest).End()
+
+	if errs != nil {
+		return "", fmt.Errorf("could not log in to auth0, error: %v", errs)
+	}
+
+	if res.StatusCode != 200 {
```

The Auth0 API is pretty finicky so it would be good to be more careful here. At least in the Kibana provider, we had a case where we were swallowing errors with 200 status codes. Even without this edge case it would be good for the user to know _why_ it's a non-200 status code (MFA reuse, invalid credentials, etc. all come under 401 for example).

https://github.com/ewilde/go-kibana/blob/4f54ca8a05c16f359cc3ffed345878f84bfe445d/http_agent.go#L42
https://github.com/ewilde/go-kibana/blob/4f54ca8a05c16f359cc3ffed345878f84bfe445d/logzio_authentication_handler.go#L48

Here's how it's handled in `go-kibana`. Basically you are looking for the presence of an `error` field in the response body.

---

There is a struct `loginResponse` in the client which had only access to the token. It was expanded and we now check the error and description from Auth0.